presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java

```diff
@@ -276,59 +281,74 @@ public IPrestoSparkTaskExecutor doCreate(
-                checkArgument(shuffleInput != null || broadcastInput != null, "Input not found for sourceFragmentId: %s", sourceFragmentId);
-                checkArgument(shuffleInput == null || broadcastInput == null, "Single remote source cannot accept both, broadcast and shuffle inputs");
+                List<PrestoSparkSerializedPage> localInput = inputs.getLocalInputs().get(sourceFragmentId.toString());
+
                 if (shuffleInput != null) {
-                    shuffleRemoteSourceInputs.add(Iterators.transform(shuffleInput, tuple -> tuple._2));
+                    checkArgument(broadcastInput == null, "single remote source is not expected to accept different kind of inputs");
+                    checkArgument(localInput == null, "single remote source is not expected to accept different kind of inputs");
+                    remoteSourceRowInputs.add(Iterators.transform(shuffleInput, tuple -> tuple._2));
+                    continue;
                 }
+
                 if (broadcastInput != null) {
+                    checkArgument(localInput == null, "single remote source is not expected to accept different kind of inputs");
                     // TODO: Enable NullifyingIterator once migrated to one task per JVM model
                     // NullifyingIterator removes element from the list upon return
                     // This allows GC to gradually reclaim memory
-                    // broadcastRemoteSourceInputs.add(getNullifyingIterator(broadcastInput.value()));
-                    broadcastRemoteSourceInputs.add(broadcastInput.value().iterator());
+                    // remoteSourcePageInputs.add(getNullifyingIterator(broadcastInput.value()));
+                    remoteSourcePageInputs.add(broadcastInput.value().iterator());
+                    continue;
                 }
+
+                if (localInput != null) {
+                    remoteSourcePageInputs.add(localInput.iterator());
+                    continue;
+                }
+
+                throw new IllegalArgumentException("Input not found for sourceFragmentId: " + sourceFragmentId);
             }
-            if (!shuffleRemoteSourceInputs.isEmpty()) {
-                shuffleInputs.put(remoteSource.getId(), Iterators.concat(shuffleRemoteSourceInputs.iterator()));
+            if (!remoteSourceRowInputs.isEmpty()) {
+                rowInputs.put(remoteSource.getId(), Iterators.concat(remoteSourceRowInputs.iterator()));
             }
-            if (!broadcastRemoteSourceInputs.isEmpty()) {
-                broadcastInputs.put(remoteSource.getId(), Iterators.concat(broadcastRemoteSourceInputs.iterator()));
+            if (!remoteSourcePageInputs.isEmpty()) {
+                pageInputs.put(remoteSource.getId(), Iterators.concat(remoteSourcePageInputs.iterator()));
             }
         }
 
+        OutputBufferMemoryManager memoryManager = new OutputBufferMemoryManager(
```

Comment:
Any reason why this block of code is now moved after the for-loop?

Reply:
I don't know why it was before the loop. It is not needed in the loop. It is usually keep the variables that are of the same scope closer together, as it is more readable.